test/unit-tests.c

```diff
@@ -325,6 +325,37 @@ test_process_command_free_passed_fds(void **state __attribute__((unused)))
     assert_int_equal(0, process_request(&vfu_ctx));
 }
 
+static void
+test_realize_ctx(void **state __attribute__((unused)))
+{
+    vfu_reg_info_t *cfg_reg;
+    vfu_reg_info_t reg_info[VFU_PCI_DEV_NUM_REGIONS + 1] = { 0 };
+    vfu_ctx_t vfu_ctx = {
+        .reg_info = reg_info,
+        .nr_regions = VFU_PCI_DEV_NUM_REGIONS + 1
+    };
+
+    assert_int_equal(0, vfu_realize_ctx(&vfu_ctx));
+    assert_true(vfu_ctx.realized);
+    cfg_reg = &vfu_ctx.reg_info[VFU_PCI_DEV_CFG_REGION_IDX];
+    assert_int_equal(VFU_REGION_FLAG_RW, cfg_reg->flags);
+    assert_int_equal(PCI_CFG_SPACE_SIZE, cfg_reg->size);
+    assert_non_null(vfu_ctx.pci.config_space);
+    assert_non_null(vfu_ctx.irqs);
+    assert_null(vfu_ctx.pci.caps);
+}
+
+static void
+test_attach_ctx(void **state __attribute__((unused)))
+{
+    vfu_ctx_t vfu_ctx = {
+        .trans = &sock_transport_ops,
+        .fd = 111
+    };
+
+    assert_int_equal(-1, vfu_attach_ctx(&vfu_ctx));
```

Comment:
We must check that `vfu_attach_ctx` calls the transport attach, so we need to set `vfu_ctx->trans->attach` to some dummy function that returns a known value and then ensure that the dummy function is called with the expected arguments and that it returns a known value.